src/jit/MemoryInl.h

```diff
@@ -1329,37 +1494,84 @@ static void emitAtomic(sljit_compiler* compiler, Instruction* instr)
 
     addr.check(compiler, operands, offset, size);
 
+#if (defined SLJIT_32BIT_ARCHITECTURE && SLJIT_32BIT_ARCHITECTURE)
+    JITArg srcValue;
     JITArg dst;
-    sljit_s32 srcReg;
+    JITArgPair srcExpectedPair;
+    sljit_s32 tmpReg;
+    sljit_s32 srcExpectedReg;
 
-#if (defined SLJIT_32BIT_ARCHITECTURE && SLJIT_32BIT_ARCHITECTURE)
     if ((operationSize & SLJIT_32) || operationSize == SLJIT_MOV32) {
-#endif /* SLJIT_32BIT_ARCHITECTURE */
-        JITArg src(operands + 1);
-        dst = JITArg(operands + 2);
-        srcReg = GET_SOURCE_REG(src.arg, instr->requiredReg(2));
-#if (defined SLJIT_32BIT_ARCHITECTURE && SLJIT_32BIT_ARCHITECTURE)
+        JITArg tmp(operands + 0);
+        JITArg srcExpected(operands + 1);
+        srcValue = JITArg(operands + 2);
+        dst = JITArg(operands + 3);
+        tmpReg = GET_SOURCE_REG(tmp.arg, instr->requiredReg(1));
+        srcExpectedReg = GET_SOURCE_REG(srcExpected.arg, instr->requiredReg(2));
     } else {
-        JITArgPair srcPair(operands + 1);
-        JITArgPair dstPair(operands + 2);
-        srcReg = GET_TARGET_REG(srcPair.arg1, instr->requiredReg(2));
-
+        JITArgPair tmpPair(operands + 0);
+        srcExpectedPair = JITArgPair(operands + 1);
+        JITArgPair srcValuePair(operands + 2);
+        JITArgPair dstPair(operands + 3);
+        tmpReg = GET_TARGET_REG(tmpPair.arg1, instr->requiredReg(1));
+        srcExpectedReg = GET_TARGET_REG(srcExpectedPair.arg1, instr->requiredReg(2));
+
+        srcValue.arg = srcValuePair.arg1;
+        srcValue.argw = srcValuePair.arg1w;
         dst.arg = dstPair.arg1;
         dst.argw = dstPair.arg1w;
         sljit_emit_op1(compiler, SLJIT_MOV, dstPair.arg2, dstPair.arg2w, SLJIT_IMM, 0);
     }
-#endif /* SLJIT_32BIT_ARCHITECTURE */
 
+    struct sljit_jump* compareFalse;
+    struct sljit_jump* compareTopFalse;
+    struct sljit_jump* storeSuccess;
     struct sljit_label* restartOnFailure = sljit_emit_label(compiler);
+
+    if (!(operationSize & SLJIT_32) && operationSize != SLJIT_MOV32) {
+        compareTopFalse = sljit_emit_cmp(compiler, SLJIT_NOT_EQUAL, SLJIT_IMM, 0, srcExpectedPair.arg2, srcExpectedPair.arg2w);
+    }
+    sljit_emit_op1(compiler, SLJIT_MOV, tmpReg, 0, srcValue.arg, srcValue.argw);
```

Comment:
This is just a note. It looks like webassembly is sensitive to the expected value upper bits (must be 0), but does not care about the srcValue upper bits. Does not make much sense.